api/build/graph.go

```diff
@@ -0,0 +1,577 @@
+	Destination int `json:"destination"`
+
+	// vela metadata
+	Status string `json:"status"`
+}
+
+// stg represents a stage's steps and some metadata for producing node/edge information
+type stg struct {
+	steps []*library.Step
+	// used for tracking stage status
+	success    int
+	running    int
+	failure    int
+	killed     int
+	canceled   int
+	skipped    int
+	startedAt  int
+	finishedAt int
+}
+
+const (
+	// clusters determine graph orientation
+	BuiltInCluster       = 2
+	PipelineCluster      = 1
+	ServiceCluster       = 0
+	GraphComplexityLimit = 1000 // arbitrary value to limit render complexity
+)
+
+// swagger:operation GET /api/v1/repos/{org}/{repo}/builds/{build}/graph builds GetBuildGraph
+//
+// Get directed a-cyclical graph for a build in the configured backend
+//
+// ---
+// produces:
+// - application/json
+// parameters:
+// - in: path
+//   name: org
+//   description: Name of the org
+//   required: true
+//   type: string
+// - in: path
+//   name: repo
+//   description: Name of the repo
+//   required: true
+//   type: string
+// - in: path
+//   name: build
+//   description: Build number
+//   required: true
+//   type: integer
+// security:
+//   - ApiKeyAuth: []
+// responses:
+//   '200':
+//     description: Successfully retrieved graph for the build
+//     schema:
+//       type: array
+//       items:
+//         "$ref": "#/definitions/Graph"
+//   '500':
+//     description: Unable to retrieve graph for the build
+//     schema:
+//       "$ref": "#/definitions/Error"
+
+// GetBuildGraph represents the API handler to capture a
+// directed a-cyclical graph for a build from the configured backend.
+func GetBuildGraph(c *gin.Context) {
+	// capture middleware values
+	b := build.Retrieve(c)
+	o := org.Retrieve(c)
+	r := repo.Retrieve(c)
+	u := user.Retrieve(c)
+	m := c.MustGet("metadata").(*types.Metadata)
+	ctx := c.Request.Context()
+
+	entry := fmt.Sprintf("%s/%d", r.GetFullName(), b.GetNumber())
+	logger := logrus.WithFields(logrus.Fields{
+		"build": b.GetNumber(),
+		"org":   o,
+		"repo":  r.GetName(),
+		"user":  u.GetName(),
+	})
+
+	baseErr := "unable to retrieve graph"
+
+	// update engine logger with API metadata
+	//
+	// https://pkg.go.dev/github.com/sirupsen/logrus?tab=doc#Entry.WithFields
+	logger.Infof("constructing graph for build %s", entry)
+
+	// retrieve the steps for the build from the step table
+	steps := []*library.Step{}
+	page := 1
+	perPage := 100
+	for page > 0 {
+		// retrieve build steps (per page) from the database
+		stepsPart, _, err := database.FromContext(c).ListStepsForBuild(b, nil, page, perPage)
+		if err != nil {
+			retErr := fmt.Errorf("unable to retrieve steps for build %s: %w", entry, err)
+
+			util.HandleError(c, http.StatusNotFound, retErr)
+
+			return
+		}
+
+		// add page of steps to list steps
+		steps = append(steps, stepsPart...)
+
+		// assume no more pages exist if under 100 results are returned
+		//
+		// nolint: gomnd // ignore magic number
```

Comment:
linter typically yells at these to for not being `//nolint:` vs `// nolint:` 🤔

Reply:
you can tell how long ago i copy pasted the code 😅 i removed them